Add support for multiple iterations

makortel opened the pull request:
#### PR description:

This PR is work-in-progress for developments towards supporting multiple iterations with mkFit. Code compiles, and RECO runs, but there are still some things to do (not necessarily in this order, or in this PR)

Affecting physics or timing
- [x] test that multi-iteration technically works by replicating the initialStep-mkFit in highPtTripletStep (without hit masks)
- [x] make the highPtTripletStep test more realistic by using highPtTripletStep seeds
- [x] add support for hit masks when the interface is ready in mkFit
- [x] add support for different strip cluster charge cut values per iteration
- [ ] ~~copy code generating `IterationConfig` into CMSSW~~
- [x] allow different `IterationConfig` objects per iteration as they are defined in mkFit's `CMS-2017.cc`
- [x] look into creating `IterationConfig` from `MkFitProducer` configuration
- [x] RECO configuration file to run (similar to #6)
  - [x] mechanism to enable mkFit for all (supported) iterations or some subset (including individual iterations)
  - [x] support for tracking ntuple output
-  [x] option to use tbb::task_arena to limit concurrency within mkFit to 1; this should enable proper time accounting by framework timers or FastTimerService in multithreaded jobs
- [x] support for patching iteration configuration via JSON file (using the mechanism in trackreco/mkFit#305
- [x] change of #13


Technical cleanup
- [ ] ~~copy code generating `TrackerInfo` into CMSSW̃~~
- [ ] look into creating `TrackerInfo` from tracker geometry information in EventSetup
- [ ] incorporate `LayerNumberConverter` into CMSSW
- [ ] remove the need to `#define USE_MATRIPLEX` in CMSSW
- [ ] address `warning: ignoring #pragma omp simd` warnings caused by `#include "mkFit/SteeringParams.h"`
- [ ] remove remaining writes to global variables
   - [ ] `mkfit::MkBuilderWrapper::populate()`
   - [ ] `mkfit::ConfigWrapper::initializeForCMSSW()`
   - [ ] `mkfit::ConfigWrapper::setNTotalLayers()`

#### PR validation:

Code builds, matrix workflow 11634.7 step3 runs.

makortel commented:
Stack trace of the crash
```
#3  0x00007f02bceef444 in sig_dostack_then_abort () from /cvmfs/cms.cern.ch/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_2_0/lib/slc7_amd64_gcc900/pluginFWCoreServicesPlugins.so
#4  <signal handler called>
#5  mkfit::TrackerInfo::outer_barrel_layer (this=<optimized out>) at ../TrackerInfo.h:229
#6  _INTERNAL64cecdcd::(anonymous namespace)::PartitionSeeds0::{lambda(mkfit::TrackerInfo const&, std::vector<mkfit::Track, std::allocator<mkfit::Track> > const&, mkfit::EventOfHits const&, mkfit::IterationSeedPartition&)#1}::operator()(mkfit::TrackerInfo const&, std::vector<mkfit::Track, std::allocator<mkfit::Track> > const&, mkfit::EventOfHits const&, mkfit::IterationSeedPartition&) const (this=<optimized out>, trk_info=..., in_seeds=..., eoh=..., part=...) at CMS-2017.cc:132
#7  std::_Function_handler<void (mkfit::TrackerInfo const&, std::vector<mkfit::Track, std::allocator<mkfit::Track> > const&, mkfit::EventOfHits const&, mkfit::IterationSeedPartition&), _INTERNAL64cecdcd::(anonymous namespace)::PartitionSeeds0::{lambda(mkfit::TrackerInfo const&, std::vector<mkfit::Track, std::allocator<mkfit::Track> > const&, mkfit::EventOfHits const&, mkfit::IterationSeedPartition&)#1}>::_M_invoke(std::_Any_data const&, mkfit::TrackerInfo const&, std::vector<mkfit::Track, std::allocator<mkfit::Track> > const&, mkfit::EventOfHits const&, mkfit::IterationSeedPartition&) (__functor=..., __args=..., __args=..., __args=..., __args=...) at /cvmfs/cms.cern.ch/slc7_amd64_gcc900/external/gcc/9.3.0/bin/../include/c++/9.3.0/bits/std_function.h:300
#8  0x00007f0291976bf8 in std::function<void (mkfit::TrackerInfo const&, std::vector<mkfit::Track, std::allocator<mkfit::Track> > const&, mkfit::EventOfHits const&, mkfit::IterationSeedPartition&)>::operator()(mkfit::TrackerInfo const&, std::vector<mkfit::Track, std::allocator<mkfit::Track> > const&, mkfit::EventOfHits const&, mkfit::IterationSeedPartition&) const (this=<optimized out>, __args=..., __args=..., __args=..., __args=...) at /cvmfs/cms.cern.ch/slc7_amd64_gcc900/external/gcc/9.3.0/bin/../include/c++/9.3.0/bits/std_function.h:688
#9  mkfit::MkBuilder::import_seeds(std::vector<mkfit::Track, std::allocator<mkfit::Track> > const&, std::function<void (mkfit::Track const&)>) (this=0x7f017b9fcf40, in_seeds=..., insert_seed=<error reading variable: Cannot access memory at address 0x0>) at MkBuilder.cc:230
#10 0x00007f029197733e in mkfit::MkBuilder::find_tracks_load_seeds (this=0x7f017b9fcf40, in_seeds=...) at MkBuilder.cc:1584
#11 0x00007f02919ae761 in mkfit::run_OneIteration (itconf=..., eoh=..., builder=..., seeds=..., out_tracks=..., do_seed_clean=3, do_backward_fit=true, do_remove_duplicates=true) at buildtestMPlex.cc:479
#12 0x00007f0291abc600 in tbb::interface7::internal::delegated_function<MkFitProducer::produce(edm::StreamID, edm::Event&, edm::EventSetup const&) const::{lambda()#2} const, void>::operator()() const () from .../CMSSW_11_2_0/lib/slc7_amd64_gcc900/pluginRecoTrackerMkFitPlugins.so
#13 0x00007f02c9aeed45 in tbb::interface7::internal::isolate_within_arena (d=..., isolation=<optimized out>) at ../../src/tbb/arena.cpp:1199
#14 0x00007f0291abcade in MkFitProducer::produce(edm::StreamID, edm::Event&, edm::EventSetup const&) const () from .../CMSSW_11_2_0/lib/slc7_amd64_gcc900/pluginRecoTrackerMkFitPlugins.so
```

makortel commented:
With d10a521839db0e5eafafe287f50c517fbc5f7ceb and https://github.com/trackreco/mkFit/pull/297 the mkFit itself technically runs, but the downstream modules fail to find their input because of remaining `#ifdef TODO` in `MkFitOutputConverter`.

makortel commented:
With cba981f266f69520d856033352f1d26c5faec9db RECO runs, and with https://github.com/trackreco/mkFit/pull/298 results of initialStep are pretty close (but not identical) to `stable-2020` + `CMSSW_11_2_0_mkFit_X` (plots made with 1k events of ttbar+PU)
https://mkortela.web.cern.ch/mkortela/tracking/mkfit/CMSSW_11_2_0/multiIter_v2/ttbar_pu50_initialStep.html

makortel commented:
In d8eb157 mkFit is being used in highPtTripletStep with initialStep steering parameters and without hit mask. The MTV plots are here
https://mkortela.web.cern.ch/mkortela/tracking/mkfit/CMSSW_11_2_0/multiIter_v4/
(the blue is mkFit tracks in initialStep, and CKF tracks in highPtTripletStep)

slava77 commented:
@makortel 
we talked about making a new branch to virtually "freeze" this one.

should we make CMSSW_11_2_2_mkFit_X and imply that we advance to this more recent release?

makortel commented:
Cleaned commit history by squashing, and added support for
- different `IterationConfig` objects per iteration as defined in `CMS-2017.cc`
- cluster masking
- different strip cluster charge cut value per iteration

makortel commented:
This update includes the following changes
* configuration file for (tracking-only) RECO step (partially from #6)
* mechanism to switch individual iterations, set of iterations, or all (supported) iterations to mkFit
* option to use `tbb::task_arena` to limit concurrency within mkFit to 1; this should enable proper time accounting by framework timers or FastTimerService in multithreaded jobs
* support for tracking ntuple output
* support for patching iteration configuration via JSON file (using the mechanism in https://github.com/trackreco/mkFit/pull/305

I think this branch is now feature-wise complete (for what is needed now).

makortel commented:
This update includes
- trackingNtuple stores SimHit information also for pileup (or at least doesn't complain)
- enable resolution histograms in MTV for all iterations in standalone and trackingOnly mode

makortel commented:
Now includes also #13

slava77 commented on RecoTracker/MkFit/plugins/MkFitHitConverter.cc:
```diff
@@ -185,19 +220,37 @@ void MkFitHitConverter::convertHits(const HitCollection& hits,
       err.At(1, 2) = gerr.czy();
 
+      auto clusterRef = hit.firstClusterRef();
+      if UNLIKELY (clusterRef.id() != clusterID) {
```
does `[UN]UNLIKELY` matter anymore?
I've seen that on more recent (10 years perhaps) the hardware wouldn't care because the choice is dynamic and it's just as effective without the hints.

makortel replied:
Good question, I don't know.

slava77 replied:
while looking at C++20 features I learned of `[[likely]]` and `[[unlikely]]`
http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0479r5.html
there are supported in gcc 9, apparently.

So, it looks like at least formally this feature is still alive

slava77 commented on RecoTracker/MkFit/test/reco_harvest_cfg.py:
```diff
@@ -0,0 +1,107 @@
+
+from Configuration.Eras.Era_Run3_cff import Run3
+
+process = cms.Process('HARVESTING',Run3)
```
is there a way to get this directly from cmsDriver?

makortel replied:
The harvesting configuration file is directly from 11634.1, so yes, it can be obtained directly from `cmsDriver`.

In principle `harvestTrackValidationPlots.py` script would also do the job.

Having an explicit file was sort of a personal preference for a CMSSW-version-wise stable setup. (and what we have in 10_4_0)

slava77 commented:
@makortel 
at a rather cursory look, I do not see any significant problems that would prevent merging this for our needs. I wouldn't even expect much more in comments in the cms-sw/cmssw review
- perhaps to get the hits definition out of the InitialStep\*py. 
- we discussed already a bit the issue of the configuration still packaged in mkFit; perhaps an example json override would help to show a way for tuning withouth rebuilding the external.

makortel commented:
> at a rather cursory look, I do not see any significant problems that would prevent merging this for our needs. I wouldn't even expect much more in comments in the cms-sw/cmssw review

Thanks for the review!

> * perhaps to get the hits definition out of the InitialStep*py.

I'm open for suggestions of better place. If we move `mkFitHits`, I'd also move `mkFitGeometryESProducer` to the same place. `iterativeTk_cff.py` directly?

On the other hand, (in absence of better way) I'd like include the timing of `mkFitHits` as part of (e.g.) InitialStep.

> * we discussed already a bit the issue of the configuration still packaged in mkFit; perhaps an example json override would help to show a way for tuning withouth rebuilding the external.

There is an example JSON file in the mkFit repo https://github.com/trackreco/mkFit/blob/devel/mkFit/config-parse/test.json. I can copy it here too if that would make it easier to find.

makortel commented:
> > * we discussed already a bit the issue of the configuration still packaged in mkFit; perhaps an example json override would help to show a way for tuning withouth rebuilding the external.
> 
> There is an example JSON file in the mkFit repo https://github.com/trackreco/mkFit/blob/devel/mkFit/config-parse/test.json. I can copy it here too if that would make it easier to find.

I added an example file. I think the rest can be addressed later. @slava77 if you agree, I'd then rebase-merge this PR.
